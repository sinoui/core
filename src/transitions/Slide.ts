import Slide from '@sinoui/core/Slide';

export default Slide;
