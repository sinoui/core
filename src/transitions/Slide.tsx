import * as React from 'react';
import { transitions } from '@sinoui/theme';
import createStyleTransition from './createStyleTransition';
import { BaseTransitionCompProps } from './type';

const { duration } = transitions;

export interface SlidePropsType extends BaseTransitionCompProps {
  appear?: boolean;
  in?: boolean;
  /**
   * 过渡效果的方向
   */
  direction?: 'up' | 'down';
  /**
   * 子节点
   */
  children?: React.ReactNode;
  /**
   * 过渡时长。
   */
  timeout?:
    | number
    | {
        enter: number;
        exit: number;
      };
  /**
   * 指定过渡时应用的css类名。
   */
  transitionClasses?:
    | string
    | {
        appear?: string;
        appearActive?: string;
        enter?: string;
        enterActive?: string;
        exit?: string;
        exitActive?: string;
      };
  /**
   * 过渡效果的缓动函数
   */
  easing?:
    | string
    | {
        enter: string;
        exit: string;
      };
}

// Slide效果默认的缓动函数
const SLIDE_EASING = {
  enter: 'cubic-bezier(0.23, 1, 0.32, 1)',
  exit: 'cubic-bezier(0.23, 1, 0.32, 1)',
};

// Slide效果默认的过渡时长
const TRANSITION_DURATION = {
  enter: duration.enteringScreen,
  exit: duration.leavingScreen,
};

/**
 * 水平滑动效果。内部使用
 * [react-transition-group/CSSTransition](https://reactcommunity.org/react-transition-group/#CSSTransition)实现过渡效果。
 */
const Slide: React.ComponentType<SlidePropsType | any> = createStyleTransition({
  enterStyle: ({ props }) => ({
    transform:
      props.direction === 'up' ? 'translate(0px, 100%)' : 'translateY(0px)',
    opacity: 0.01,
  }),
  enterToStyle: ({ props }) => ({
    transform:
      props.direction === 'down' ? 'translate(0px, 100%)' : 'translateY(0px)',
    opacity: 1,
  }),
  exitStyle: ({ props }) => ({
    opacity: 1,
    transform:
      props.direction === 'up'
        ? 'translate3d(0px, 0px, 0px)'
        : 'translate(0px, 100%)',
  }),
  exitToStyle: ({ props }) => ({
    transform:
      props.direction === 'down'
        ? 'translate3d(0px, 0px, 0px)'
        : 'translate(0px, 100%)',
    opacity: 0.01,
  }),
  transitionProps: ['transform', 'opacity'],
  easing: SLIDE_EASING,
  duration: TRANSITION_DURATION,
});

export default Slide;
