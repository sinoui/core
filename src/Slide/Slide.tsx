/* eslint-disable no-param-reassign */
import React from 'react';
import { Transition } from 'react-transition-group';
import transitions, { duration } from '@sinoui/theme/transitions';
import type { TransitionStatus } from 'react-transition-group/Transition';
import BaseTransitionProps from '../transitions/BaseTransitionProps';
import useMultiRefs from '../utils/useMultiRefs';

type Direction = 'down' | 'up' | 'left' | 'right';

type Props = BaseTransitionProps & {
  /**
   * 指定元素平移出现的方向。默认为`down`。
   */
  direction?: Direction;
};

const defaultTimeout = {
  enter: duration.enteringScreen,
  exit: duration.leavingScreen,
};

export function getTranslate(node: HTMLElement, direction: Direction) {
  const rect = node.getBoundingClientRect();

  switch (direction) {
    case 'left':
      return `translateX(${window.innerWidth - rect.left}px)`;
    case 'right':
      return `translateX(-${rect.left + rect.width}px)`;
    case 'down':
      return `translateY(-${rect.top + rect.height}px)`;
    case 'up':
    default:
      return `translateY(${window.innerHeight - rect.top}px)`;
  }
}

/**
 * 重排元素。
 *
 * 调用此方法，浏览器会立马应用新的css样式。
 *
 * **注意**：此方法会引起全局重排，所以慎用。
 *
 * @param node 元素
 */
const reflow = (node: HTMLElement) => node.scrollTop;

/**
 * 从屏幕边缘平移滑动的过渡效果。
 */
const Slide = React.forwardRef<HTMLElement, Props>(function Slide(props, ref) {
  const {
    direction = 'down',
    children,
    in: inProp,
    timeout = defaultTimeout,
    onEnter,
    onEntering,
    onExit,
    onExited,
    ...rest
  } = props;

  const handleRef = useMultiRefs(ref, (children as any).ref);

  /**
   * 进入过渡的初始化回调函数。在初始化步骤需要将元素设置到屏幕外部。
   *
   * @param node 节点
   * @param isAppearing 指示进场动画是否触发在初始挂载时
   */
  const handleEnter = (node: HTMLElement, isAppearing: boolean) => {
    node.style.transform = getTranslate(node, direction);
    reflow(node); // 修复初始化隐藏到展现时无过渡的缺陷
    if (onEnter) {
      onEnter(node, isAppearing);
    }
  };

  /**
   * 进入过渡的开始回调函数。在开始步骤中设置过渡效果，并让元素出现在之前的位置。
   *
   * @param node 节点
   * @param isAppearing 指示进场动画是否触发在初始挂载时
   */
  const handleEntering = (node: HTMLElement, isAppearinging: boolean) => {
    node.style.transition = transitions.create('transform', {
      duration: typeof timeout === 'number' ? timeout : timeout.enter,
      easing: transitions.easing.easeOut,
    });
    node.style.transform = 'none';
    if (onEntering) {
      onEntering(node, isAppearinging);
    }
  };

  /**
   * 退出过渡的回调函数。需要将元素设置到屏幕之外，并设置过渡效果。
   *
   * @param node 节点
   */
  const handleExit = (node: HTMLElement) => {
    node.style.transition = transitions.create('transform', {
      duration: typeof timeout === 'number' ? timeout : timeout.exit,
      easing: transitions.easing.sharp,
    });
    node.style.transform = getTranslate(node, direction);
    if (onExit) {
      onExit(node);
    }
  };

  /**
   * 退出过渡的完成回调函数。需要在过渡结束时清除过渡效果。
   * @param node 节点
   */
  const handleExited = (node: HTMLElement) => {
    node.style.transition = 'none';
    node.style.transform = 'none';
    if (onExited) {
      onExited(node);
    }
  };

  return (
    <Transition
      appear
      in={inProp}
      timeout={timeout}
      onEnter={handleEnter}
      onEntering={handleEntering}
      onExit={handleExit}
      onExited={handleExited}
      {...rest}
    >
      {(status: TransitionStatus, childProps: any) =>
        React.cloneElement(children, {
          ref: handleRef,
          ...childProps,
          style: {
            visibility: status === 'exited' && !inProp ? 'hidden' : undefined,
            ...childProps.style,
            ...children.props.style,
          },
        })
      }
    </Transition>
  );
});

export default Slide;
