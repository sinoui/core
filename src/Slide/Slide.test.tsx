import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { transitions } from '@sinoui/theme';
import Slide, { getTranslate } from './Slide';

afterEach(cleanup);

jest.useFakeTimers();

it('初始出现', () => {
  const { getByTestId } = render(
    <Slide in>
      <div data-testid="slide">测试</div>
    </Slide>,
  );

  expect(getByTestId('slide')).not.toHaveStyle('visibility:hidden');
  expect(getByTestId('slide')).toHaveStyle('transform: none');
  expect(getByTestId('slide')).toHaveStyle(
    `transition: ${transitions.create('transform', {
      duration: transitions.duration.enteringScreen,
      easing: transitions.easing.easeOut,
    })}`,
  );
});

it('初始隐藏', () => {
  const { getByTestId } = render(
    <Slide in={false}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );

  expect(getByTestId('slide')).toHaveStyle('visibility:hidden');
});

it('显示 -> 隐藏', () => {
  const { getByTestId, rerender } = render(
    <Slide in>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  Object.defineProperty(getByTestId('slide'), 'getBoundingClientRect', {
    configurable: true,
    value: () => ({
      top: 20,
      height: 20,
    }),
  });

  rerender(
    <Slide in={false}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  expect(getByTestId('slide').style.transform).toBe('translateY(-40px)');
  expect(getByTestId('slide')).toHaveStyle(
    `transition: ${transitions.create('transform', {
      duration: transitions.duration.leavingScreen,
      easing: transitions.easing.sharp,
    })}`,
  );

  jest.runAllTimers();
  expect(getByTestId('slide')).toHaveStyle('transform: none');
  expect(getByTestId('slide')).toHaveStyle('transition: none');
});

it('隐藏 -> 显示', () => {
  const onEnter = jest
    .fn()
    .mockImplementation((node) =>
      expect(node).toHaveStyle('transform: translateY(-40px)'),
    );

  const { getByTestId, rerender } = render(
    <Slide in={false} onEnter={onEnter}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  Object.defineProperty(getByTestId('slide'), 'getBoundingClientRect', {
    configurable: true,
    value: () => ({
      top: 20,
      height: 20,
    }),
  });
  rerender(
    <Slide in onEnter={onEnter}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );

  expect(onEnter).toBeCalled();
  expect(getByTestId('slide')).toHaveStyle('transform: none');
  expect(getByTestId('slide')).toHaveStyle(
    `transition: ${transitions.create('transform', {
      duration: transitions.duration.enteringScreen,
      easing: transitions.easing.easeOut,
    })}`,
  );
});

it('onEnter', () => {
  const onEnter = jest.fn();
  render(
    <Slide in onEnter={onEnter}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  expect(onEnter).toBeCalled();
});

it('onEntering', () => {
  const onEntering = jest.fn();
  render(
    <Slide in onEntering={onEntering}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  expect(onEntering).toBeCalled();
});

it('onExit', () => {
  const onExit = jest.fn();
  const { rerender } = render(
    <Slide in onExit={onExit}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  rerender(
    <Slide in={false} onExit={onExit}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  expect(onExit).toBeCalled();
});

it('onExited', () => {
  const onExited = jest.fn();
  const { rerender } = render(
    <Slide in onExited={onExited}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  rerender(
    <Slide in={false} onExited={onExited}>
      <div data-testid="slide">测试</div>
    </Slide>,
  );
  jest.runAllTimers();
  expect(onExited).toBeCalled();
});

describe('direcction', () => {
  it('left', () => {
    Object.defineProperty(window, 'innerWidth', {
      value: 200,
      configurable: true,
    });
    const node = document.createElement('div');
    Object.defineProperty(node, 'getBoundingClientRect', {
      value: () => ({ left: 20 }),
      configurable: true,
    });

    expect(getTranslate(node, 'left')).toBe('translateX(180px)');
  });

  it('right', () => {
    const node = document.createElement('div');
    Object.defineProperty(node, 'getBoundingClientRect', {
      value: () => ({ left: 20, width: 100 }),
      configurable: true,
    });

    expect(getTranslate(node, 'right')).toBe('translateX(-120px)');
  });

  it('down', () => {
    const node = document.createElement('div');
    Object.defineProperty(node, 'getBoundingClientRect', {
      value: () => ({ top: 20, height: 30 }),
      configurable: true,
    });

    expect(getTranslate(node, 'down')).toBe('translateY(-50px)');
  });

  it('up', () => {
    Object.defineProperty(window, 'innerHeight', {
      value: 300,
      configurable: true,
    });
    const node = document.createElement('div');
    Object.defineProperty(node, 'getBoundingClientRect', {
      value: () => ({ top: 90 }),
      configurable: true,
    });

    expect(getTranslate(node, 'up')).toBe('translateY(210px)');
  });
});

it('ref', () => {
  const outerRef = React.createRef<HTMLElement>();
  const innerRef = React.createRef<HTMLDivElement>();
  const { getByTestId } = render(
    <Slide in data-testid="slide" ref={outerRef}>
      <div ref={innerRef}>测试</div>
    </Slide>,
  );

  expect(outerRef.current).toBe(getByTestId('slide'));
  expect(innerRef.current).toBe(getByTestId('slide'));
});

it('style', () => {
  const { getByTestId } = render(
    <Slide
      in
      data-testid="slide"
      style={{
        color: 'red',
      }}
    >
      <div style={{ backgroundColor: 'yellow' }}>测试</div>
    </Slide>,
  );

  expect(getByTestId('slide')).toHaveStyle(
    'color: red; background-color: yellow',
  );
});

it('子元素属性', () => {
  const { getByTestId } = render(
    <Slide in data-testid="slide">
      <div className="slide">测试</div>
    </Slide>,
  );

  expect(getByTestId('slide')).toHaveClass('slide');
});
