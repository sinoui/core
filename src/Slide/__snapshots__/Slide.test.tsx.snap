// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 展现 1`] = `
<div
  style={
    Object {
      "visibility": undefined,
    }
  }
>
  测试
</div>
`;
